File: app/src/main/java/com/kunfei/bookshelf/widget/page/TxtChapter.java
package com.kunfei.bookshelf.widget.page;

import java.util.ArrayList;
import java.util.List;

/**
 * 章节
 */

public class TxtChapter {
    private int position;
    private List<TxtPage> txtPageList = new ArrayList<>();
    private List<Integer> txtPageLengthList = new ArrayList<>();
    private List<Integer> paragraphLengthList = new ArrayList<>();
    private Status status = Status.LOADING;
    private String msg;

    TxtChapter(int position) {
        this.position = position;
    }

    public int getPosition() {
        return position;
    }

    List<TxtPage> getTxtPageList() {
        return txtPageList;
    }

    void addPage(TxtPage txtPage) {
        txtPageList.add(txtPage);
    }

    int getPageSize() {
        return txtPageList.size();
    }

    TxtPage getPage(int page) {
        if (!txtPageList.isEmpty()) {
            return txtPageList.get(Math.max(0, Math.min(page, txtPageList.size() - 1)));
        }
        return null;
    }

    Status getStatus() {
        return status;
    }

    void setStatus(Status mStatus) {
        this.status = mStatus;
    }

    public String getMsg() {
        return msg;
    }

    public void setMsg(String msg) {
        this.msg = msg;
    }

    int getPageLength(int position) {
        if (txtPageLengthList != null && position >= 0 && position < txtPageLengthList.size()) {
            return txtPageLengthList.get(position);
        }
        return -1;
    }

    void addTxtPageLength(int length) {
        txtPageLengthList.add(length);
    }

    List<Integer> getTxtPageLengthList() {
        return txtPageLengthList;
    }

    List<Integer> getParagraphLengthList() {
        return paragraphLengthList;
    }

    void addParagraphLength(int length) {
        paragraphLengthList.add(length);
    }

    int getParagraphIndex(int length) {
        for (int i = 0; i < paragraphLengthList.size(); i++) {
            if ((i == 0 || paragraphLengthList.get(i - 1) < length) && length <= paragraphLengthList.get(i)) {
                return i;
            }
        }
        return -1;
    }

    public enum Status {
        LOADING, FINISH, ERROR, EMPTY, CATEGORY_EMPTY, CHANGE_SOURCE,
    }
}

File: app/src/main/java/com/kunfei/bookshelf/widget/page/TxtChapter.kt
package com.kunfei.bookshelf.widget.page

import java.util.*
import kotlin.math.max
import kotlin.math.min

/**
 * 章节
 */

class TxtChapter internal constructor(val position: Int) {
    val txtPageList = ArrayList<TxtPage>()
    val txtPageLengthList = ArrayList<Int>()
    val paragraphLengthList = ArrayList<Int>()
    var status = Status.LOADING
    var msg: String? = null

    val pageSize: Int
        get() = txtPageList.size

    fun addPage(txtPage: TxtPage) {
        txtPageList.add(txtPage)
    }

    fun getPage(page: Int): TxtPage? {
        return if (txtPageList.isNotEmpty()) {
            txtPageList[max(0, min(page, txtPageList.size - 1))]
        } else null
    }

    fun getPageLength(position: Int): Int {
        return if (position >= 0 && position < txtPageLengthList.size) {
            txtPageLengthList[position]
        } else -1
    }

    fun addTxtPageLength(length: Int) {
        txtPageLengthList.add(length)
    }

    fun addParagraphLength(length: Int) {
        paragraphLengthList.add(length)
    }

    fun getParagraphIndex(length: Int): Int {
        for (i in paragraphLengthList.indices) {
            if ((i == 0 || paragraphLengthList[i - 1] < length) && length <= paragraphLengthList[i]) {
                return i
            }
        }
        return -1
    }

    enum class Status {
        LOADING, FINISH, ERROR, EMPTY, CATEGORY_EMPTY, CHANGE_SOURCE
    }
}
